rfcs/0154-Migrate-taskcluster-to-postgres.md

```diff
@@ -0,0 +1,161 @@
+an queues)
+* create a worker type
+* resolve artifact
+* call the worker manager register endpoint
+* call often used endpoints like getTasks, getTaskGroup, getTaskByIndexRoute, etc.
+
+## Schema Migration
+
+Taskcluster will handle schema migrations and will have
+all migrations tested with sample data to ensure we get the expected schema. In
+addition, there will be one schema for the whole monorepo, so that schema
+updates happen atomically across all services.
+
+## Backups and Restores
+
+Teams operating Taskcluster will rely on the cloud provider's backup system to
+handle backups and restores.
+
+## DB Version Upgrades Compatibility
+
+When deploying a new version of Taskcluster, any necessary database upgrades
+*must* be applied before any services are upgraded.  Taskcluster will maintain
+the invariant that services expecting database version S can interoperate with a
+database at version V as long as V >= S.
+
+This invariant is maintained through careful attention to the definitions of the
+stored procedures (user-defined functions). These stored procedures are changed
+as necessary during the upgrade process, such that a procedure operates
+identically before and after the ugprade: identical arguments, identical return
+type, and correct behavior.
+
+_Example 1:_ If an upgrade factors a single table into two tables, then a
+procedure which previously queried from the single table would be rewritten
+during the ugprade to perform a join over the two new tables, returning
+appropriate results in the same format as before the upgrade.
+
+_Example 2:_ If you have a query that returns column foo, but in the next
+version your code wants to know the value of column bar as well, then a new
+stored procedure will need to be introduced to return both columns and stop
+using the existing stored procedure that returned the single column. That new
+stored procedure is then deployed before the code that uses it is deployed.
+
+A consequence of this design is that "procedures are forever" -- an upgrade can
+never delete a stored procedure. At worst, when a feature is removed, a stored
+procedure can be rewritten to return an empty result or perform no action.
+
+## Ad-hoc Queries
+
+Direct SQL access to the database is *not allowed*. Taskcluster will allow
+ad-hoc read-only queries on the data-set via stored procedures with access
+controlled by Postgres permissions. This feature will most likely be done after
+step 2 of the transition.
+
+## Permissions
+
+Taskcluster will manage permissions to tables/schemas and deployers will manage
+user accounts. The deployment will have an "admin" postgres user/password
+(configured in Kubernetes), and on install/upgrade we'll use the admin user to
+create a non-admin user for each service, with appropriate GRANTs for that
+service's access. Deployers of Taskcluster will pick the passwords for all the
+non-admin users (configured in Kubernetes). It's up to the deployer to create
```

Comment:
By "configured in Kubernetes", you mean "encrypted then passed as env vars like all the other secrets we currently provide to services", right?

Reply:
Correct.